structure_tests/drivers/docker_driver.go

```diff
@@ -0,0 +1,315 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package drivers
+
+import (
+	"archive/tar"
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"strings"
+	"testing"
+
+	"github.com/GoogleCloudPlatform/runtimes-common/structure_tests/types/unversioned"
+	"github.com/fsouza/go-dockerclient"
+)
+
+type DockerDriver struct {
+	cli           docker.Client
+	originalImage string
+	currentImage  string
+	env           map[string]string
+}
+
+func NewDockerDriver(image string) Driver {
+	newCli, err := docker.NewClientFromEnv()
+	if err != nil {
+		panic(err)
+	}
+	return &DockerDriver{
+		originalImage: image,
+		currentImage:  image,
+		cli:           *newCli,
+		env:           nil,
+	}
+}
+
+func (d *DockerDriver) Setup(t *testing.T, envVars []unversioned.EnvVar, fullCommand []unversioned.Command) {
+	env := d.processEnvVars(t, envVars)
+	for _, cmd := range fullCommand {
+		d.currentImage = d.runAndCommit(t, env, cmd)
+	}
+}
+
+func (d *DockerDriver) ProcessCommand(t *testing.T, envVars []unversioned.EnvVar, fullCommand []string) (string, string, int) {
+	var env []string
+	for _, envVar := range envVars {
+		env = append(env, envVar.Key+"="+envVar.Value)
+	}
+	stdout, stderr, exitCode := d.exec(t, env, fullCommand)
+
+	if stdout != "" {
+		t.Logf("stdout: %s", stdout)
+	}
+	if stderr != "" {
+		t.Logf("stderr: %s", stderr)
+	}
+	return stdout, stderr, exitCode
+}
+
+func (d *DockerDriver) retrieveEnvVar(envVar string) string {
+	// since we're only retrieving these during processing, we can cache this on the driver
+	if d.env == nil {
+		image, err := d.cli.InspectImage(d.currentImage)
+		if err != nil {
+			return ""
+		}
+
+		// convert env to map for processing
+		imageEnv := make(map[string]string)
+		for _, varPair := range image.Config.Env {
+			pair := strings.Split(varPair, "=")
+			imageEnv[pair[0]] = pair[1]
+		}
+		d.env = imageEnv
+	}
+
+	return d.env[envVar]
+}
+
+func (d *DockerDriver) processEnvVars(t *testing.T, vars []unversioned.EnvVar) []string {
+	if len(vars) == 0 {
+		return nil
+	}
+
+	env := []string{}
+
+	for _, envVar := range vars {
+		expandedVal := os.Expand(envVar.Value, d.retrieveEnvVar)
+		env = append(env, envVar.Key+"="+expandedVal)
+	}
+	return env
+}
+
+// copies a tar archive starting at the specified path from the image, and returns
+// a tar reader which can be used to iterate through its contents and retrieve metadata
+func (d *DockerDriver) retrieveTar(t *testing.T, path string) (*tar.Reader, error) {
+	// this contains a placeholder command which does not get run, since
+	// the client doesn't allow creating a container without a command.
+	container, err := d.cli.CreateContainer(docker.CreateContainerOptions{
+		Config: &docker.Config{
+			Image:        d.currentImage,
+			Cmd:          []string{"NOOP_COMMAND_DO_NOT_RUN"},
+			AttachStdout: true,
```

Comment:
I think these can be false since we don't care about the noop?

Reply:
ah yep good call